features/telemetry/daemon/src/main/java/org/opennms/netmgt/telemetry/daemon/TelemetryMessageConsumer.java

```diff
@@ -0,0 +1,136 @@
+import org.opennms.netmgt.telemetry.ipc.TelemetryMessageLogDTO;
+import org.opennms.netmgt.telemetry.ipc.TelemetrySinkModule;
+import org.opennms.netmgt.telemetry.listeners.api.TelemetryMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.BeanWrapper;
+import org.springframework.beans.PropertyAccessorFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.config.AutowireCapableBeanFactory;
+import org.springframework.context.ApplicationContext;
+
+import javax.annotation.PostConstruct;
+import java.lang.reflect.Constructor;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+public class TelemetryMessageConsumer implements MessageConsumer<TelemetryMessage, TelemetryMessageLogDTO> {
+    private final Logger LOG = LoggerFactory.getLogger(TelemetryMessageConsumer.class);
+
+    private static final ServiceParameters EMPTY_SERVICE_PARAMETERS = new ServiceParameters(Collections.emptyMap());
+
+    @Autowired
+    private ApplicationContext applicationContext;
+
+    private final Protocol protocolDef;
+    private final TelemetrySinkModule sinkModule;
+    private final List<Adapter> adapters;
+
+    public TelemetryMessageConsumer(Protocol protocol, TelemetrySinkModule sinkModule) throws Exception {
+        this.protocolDef = Objects.requireNonNull(protocol);
+        this.sinkModule = Objects.requireNonNull(sinkModule);
+        adapters = new ArrayList<>(protocol.getAdapters().size());
+    }
+
+    @PostConstruct
+    public void setUp() throws Exception {
+        // Pre-emptively instantiate the adapters
+        for (org.opennms.netmgt.telemetry.config.model.Adapter adapterDef : protocolDef.getAdapters()) {
+            try {
+                adapters.add(buildAdapter(adapterDef));
+            } catch (Exception e) {
+                throw new Exception("Failed to create adapter from definition: " + adapterDef, e);
+            }
+        }
+    }
+
+    @Override
+    public void handleMessage(TelemetryMessageLogDTO messageLog) {
+        try(Logging.MDCCloseable mdc = Logging.withPrefixCloseable(Telemetryd.LOG_PREFIX)) {
+            LOG.trace("Received message log: {}", messageLog);
+            // Handle the message with all of the adapters
+            for (Adapter adapter : adapters) {
+                try {
+                    adapter.handleMessageLog(messageLog);
+                } catch (RuntimeException e) {
+                    LOG.warn("Adapter: {} failed to handle message log: {}. Skipping.", adapter, messageLog);
+                    continue;
+                }
+            }
+        }
+    }
+
+    private Adapter buildAdapter(org.opennms.netmgt.telemetry.config.model.Adapter adapterDef) throws Exception {
+        // Instantiate the associated class
+        final Class<?> clazz = Class.forName(adapterDef.getClassName());
```

Comment:
The same as above. We may want to throw a more meaningful exception rather than a generic "ClassNotFound"-Exception in case the adapter contains a non-existing class.